cmd/clusterctl/pkg/client/delete_test.go

```diff
@@ -0,0 +1,163 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"testing"
+
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+)
+
+func Test_clusterctlClient_Delete(t *testing.T) {
+	type fields struct {
+		client *fakeClient
+	}
+	type args struct {
+		options DeleteOptions
+	}
+	tests := []struct {
+		name          string
+		fields        fields
+		args          args
+		wantProviders []string
+		wantErr       bool
+	}{
+		{
+			name: "Delete all the providers",
+			fields: fields{
+				client: fakeClusterForDelete(),
+			},
+			args: args{
+				options: DeleteOptions{
+					Kubeconfig:           "kubeconfig",
+					ForceDeleteNamespace: false,
+					ForceDeleteCRD:       false,
+					Namespace:            "",
+					ProvidersToDelete:    nil, // nil means all the providers
+				},
+			},
+			wantProviders: nil,
+			wantErr:       false,
+		},
+		{
+			name: "Delete single provider",
+			fields: fields{
+				client: fakeClusterForDelete(),
+			},
+			args: args{
+				options: DeleteOptions{
+					Kubeconfig:           "kubeconfig",
+					ForceDeleteNamespace: false,
+					ForceDeleteCRD:       false,
+					Namespace:            "capbpk-system",
+					ProvidersToDelete:    []string{bootstrapProviderConfig.Name()},
+				},
+			},
+			wantProviders: []string{capiProviderConfig.Name()},
+			wantErr:       false,
+		},
+		{
+			name: "Delete single provider auto-detect namespace",
+			fields: fields{
+				client: fakeClusterForDelete(),
+			},
+			args: args{
+				options: DeleteOptions{
+					Kubeconfig:           "kubeconfig",
+					ForceDeleteNamespace: false,
+					ForceDeleteCRD:       false,
+					Namespace:            "", // empty namespace triggers namespace auto detection
+					ProvidersToDelete:    []string{bootstrapProviderConfig.Name()},
+				},
+			},
+			wantProviders: []string{capiProviderConfig.Name()},
+			wantErr:       false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := tt.fields.client.Delete(tt.args.options); (err != nil) != tt.wantErr {
+				t.Fatalf("error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if tt.wantErr {
+				return
+			}
+
+			proxy := tt.fields.client.clusters["kubeconfig"].Proxy()
+			gotProviders := &clusterctlv1.ProviderList{}
+
+			c, err := proxy.NewClient()
+			if err != nil {
+				t.Fatalf("failed to create client %v", err)
+			}
+
+			if err := c.List(context.TODO(), gotProviders); err != nil {
+				t.Fatalf("failed to read providers %v", err)
+			}
+
+			if len(gotProviders.Items) != len(tt.wantProviders) {
+				t.Errorf("got = %v items, want %v items", len(gotProviders.Items), len(tt.wantProviders))
+				return
+			}
+
+			for _, gotProvider := range gotProviders.Items {
+				found := false
+				for _, wantProvider := range tt.wantProviders {
+					if gotProvider.Name == wantProvider {
+						found = true
```

Comment:
break after?

Reply:
Or what if instead you create 2 `sets.String` - 1 for got, 1 for want, and then you checked if the 2 sets were equal?

Reply:
done